runtime/yaml.cpp

```diff
@@ -0,0 +1,160 @@
+#include "runtime/yaml.h"
+
+void yaml_node_to_mixed(const YAML::Node &node, mixed &data, const string &source) {
+  data.clear(); // sets data to NULL
+  if (node.IsScalar()) {
+    const string string_data(node.as<std::string>().c_str());
+    if (string_data == string("true") || string_data == string("false")) {
+      if (source[node.Mark().pos] == '"' && source[node.Mark().pos + string_data.size() + 1] == '"') {
+        data = string_data;
+      } else {
+        data = string_data == string("true");
+      }
+    } else if (string_data.is_int()) {
+      if (source[node.Mark().pos] == '"' && source[node.Mark().pos + string_data.size() + 1] == '"') {
+        data = string_data;
+      } else {
+        data = string_data.to_int();
+      }
+    } else {
+      dl::enter_critical_section();
+      auto *float_data = new double;
+      if (string_data.try_to_float(float_data)) {
+        if (source[node.Mark().pos] == '"' && source[node.Mark().pos + string_data.size() + 1] == '"') {
+          data = string_data;
+        } else {
+          data = *float_data;
+        }
+      } else {
+        data = string_data;
+      }
+      delete float_data;
+      dl::leave_critical_section();
+    }
+  } else if (node.IsSequence()) {
+    for (auto it = node.begin(); it != node.end(); ++it) {
+      mixed data_piece;
+      yaml_node_to_mixed(*it, data_piece, source);
+      data.push_back(data_piece);
+    }
+  } else if (node.IsMap()) {
+    for (const auto &it : node) {
+      mixed data_piece;
+      yaml_node_to_mixed(it.second, data_piece, source);
+      data[string(it.first.as<std::string>().c_str())] = data_piece;
+    }
+  }
+}
+
+string print_tabs(uint8_t nesting_level) {
+  string tabs;
+  for (uint8_t i = 0; i < 2 * nesting_level; i++) {
+    tabs.push_back(' ');
+  }
+  return tabs;
+}
+
+string print_key(const mixed& data_key) {
+  if (data_key.is_string()) {
+    return data_key.as_string();
+  }
+  return string(data_key.as_int()); // array can not be a key; bool and float keys are cast to int
+}
+
+void mixed_to_string(const mixed& data, string& string_data, uint8_t nesting_level = 0) {
+  if (!data.is_array()) {
+    if (data.is_null()) {
+      string_data.push_back('~');
+    } else if (data.is_string()) {
+      const string& string_data_piece = data.as_string();
+      if (string_data_piece.size() < 2
+          || (string_data_piece[0] != '"' && string_data_piece[string_data_piece.size() - 1] != '"')) {
+        string_data.push_back('"');
+        string_data.append(string_data_piece);
+        string_data.push_back('"');
+      } else {
+        string_data.append(string_data_piece);
+      }
+    } else if (data.is_int()) {
+      string_data.append(data.as_int());
+    } else if (data.is_float()) {
+      string_data.append(data.as_double());
+    } else if (data.is_bool()) {
+      string_data.append((data.as_bool()) ? "true" : "false");
+    }
+    string_data.push_back('\n');
+    return;
+  }
```

Comment:
As for me, it's better to collect the string first, and then call the `append` function once.
And I guess, this code will look better if you write it as follows:

```c++
  string buffer = "";
  if (!data.is_array()) {
    if (data.is_null()) {
      buffer = '~';
    } else if (data.is_string()) {
      const string& string_data_piece = data.as_string();
      if (string_data_piece.size() < 2 || (string_data_piece[0] != '"' && string_data_piece[string_data_piece.size() - 1] != '"')) {
        buffer = "\"" + string_data_piece + "\"";
      } else {
        buffer = string_data_piece;
      }
    } else if (data.is_int()) {
      buffer = data.as_int();
    } else if (data.is_float()) {
      buffer = data.as_double();
    } else if (data.is_bool()) {
      buffer = (data.as_bool()) ? "true" : "false");
    }
    string_data.append(buffer + '\n');
    return;
  }
```

Reply:
It seems that the ```string``` class has no overloads for ```string& operator=(const string& s)``` and ```string operator+(const string& left, const string& right)```

Reply:
Also, there is no overload for ```string operator+(const string& lhs, const string& rhs)```. ```string buffer = string("something") + string("anything")``` would not compile as the ```mixed operator+(const mixed& lhs, const mixed& rhs)``` would be called and there is no ```string operator=(const mixed&)```

This can be solved by an implicit cast to ```string```, but I think this is not a good idea

Reply:
Yes, I completely forgot that kphp uses it's own strings.
I'm afraid if you rewrite my code with `kphp strings`, then readability will not only improve, but even worsen 😅
Therefore, you are absolutely right, I apologize for the disinformation!